fish-rust/src/function.rs

```diff
@@ -0,0 +1,711 @@
+use crate::env::{EnvStack, Environment};
+use crate::event::{self, EventDescription};
+use crate::ffi::{self, parser_t, Repin};
+use crate::global_safety::RelaxedAtomicBool;
+use crate::parse_tree::{NodeRef, ParsedSourceRefFFI};
+use crate::parser_keywords::parser_keywords_is_reserved;
+use crate::wchar::{wstr, WString, L};
+use crate::wchar_ext::WExt;
+use crate::wchar_ffi::wcstring_list_ffi_t;
+use crate::wchar_ffi::{AsWstr, WCharFromFFI, WCharToFFI};
+use crate::wutil::{dir_iter::DirIter, gettext::wgettext_expr, sprintf};
+use cxx::{CxxWString, UniquePtr};
+use once_cell::sync::Lazy;
+use std::collections::{HashMap, HashSet};
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+
+#[derive(Clone)]
+pub struct FunctionProperties {
+    /// Reference to the node, along with the parsed source.
+    pub func_node: NodeRef<ast::BlockStatement>,
+
+    /// List of all named arguments for this function.
+    pub named_arguments: Vec<WString>,
+
+    /// Description of the function.
+    pub description: WString,
+
+    /// Mapping of all variables that were inherited from the function definition scope to their
+    /// values.
+    pub inherit_vars: HashMap<WString, Vec<WString>>,
+
+    /// Set to true if invoking this function shadows the variables of the underlying function.
+    pub shadow_scope: bool,
+
+    /// Whether the function was autoloaded.
+    /// This is the only field which is mutated after the properties are created.
+    pub is_autoload: RelaxedAtomicBool,
+
+    /// The file from which the function was created, or None if not from a file.
+    pub definition_file: Option<FilenameRef>,
+
+    /// Whether the function was copied.
+    pub is_copy: bool,
+
+    /// The file from which the function was copied, or nullptr if not from a file.
+    pub copy_definition_file: Option<FilenameRef>,
+
+    /// The line number where the specified function was copied.
+    pub copy_definition_lineno: i32,
+}
+
+pub type FunctionPropertiesRef = Arc<FunctionProperties>;
+
+/// Type wrapping up the set of all functions.
+/// There's only one of these; it's managed by a lock.
+struct FunctionSet {
+    /// The map of all functions by name.
+    funcs: HashMap<WString, FunctionPropertiesRef>,
+
+    /// Tombstones for functions that should no longer be autoloaded.
+    autoload_tombstones: HashSet<WString>,
+
+    /// The autoloader for our functions.
+    autoloader: cxx::UniquePtr<ffi::autoload_t>,
+}
+
+impl FunctionSet {
+    /// Remove a function.
+    /// \return true if successful, false if it doesn't exist.
+    fn remove(&mut self, name: &wstr) -> bool {
+        if self.funcs.remove(name).is_some() {
+            event::remove_function_handlers(name);
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Get the properties for a function, or None if none.
+    fn get_props(&self, name: &wstr) -> Option<FunctionPropertiesRef> {
+        self.funcs.get(name).cloned()
```

Comment:
Am I the only one that would find all our rust code to imminently more readable if we got rid of the type aliases? Seeing `clone()` used everywhere gives me instinctual panic, then I realize the type we are returning is an `Arc`.

Reply:
I've removed this type.